### PR TITLE
Added some tests for checking set_priority

### PR DESCRIPTION
**Pull request by rgundi**
Added some scenarios for threads to test the usage of
set_priority.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

### CODE REVIEW

**Comment by nashif** [comment]
recheck

File: tests/kernel/threads/usage/testcase.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+tests:
+  kernel.threads:
+    arch_exclude: nios2
```
Comment (nashif) [inline tests/kernel/threads/usage/testcase.yaml]: why is this not working on nios2?

Comment (rgundi) [reply tests/kernel/threads/usage/testcase.yaml]: Well.. i copied this from tests/kernel/threads/lifecycle/lifecycle_api/testcase.yaml. So, not sure why this was excluded there.

File: tests/kernel/threads/usage/src/main.c [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+#include <ztest.h>
+extern void test_thread1(void);
+extern void test_thread2(void);
+
+__kernel struct k_thread tdata;
+#define STACK_SIZE (2048 + CONFIG_TEST_EXTRA_STACKSIZE)
+K_THREAD_STACK_DEFINE(tstack, STACK_SIZE);
+
```
Comment (ramakrishnapallala) [inline tests/kernel/threads/usage/src/main.c]: what is the use of the above thread strutcs and stack? any way you defining them again in task.c

File: tests/kernel/threads/usage/src/task.c [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+	k_sem_give(&sem_thread2);
+	k_sem_take(&sem_thread1, K_FOREVER);
+	zassert_equal(thread2_data, THREAD2_PRIO + 2,
+		      "Expected priority to be changed to %d, not %d\n",
+		      THREAD2_PRIO + 2, thread2_data);
+
+	/* Raise the priority of thread2 */
+	k_thread_priority_set(thread2_id, THREAD2_PRIO - 2);
+	k_sem_give(&sem_thread2);
+	k_sem_take(&sem_thread1, K_FOREVER);
```
Comment (ramakrishnapallala) [inline tests/kernel/threads/usage/src/task.c]: do we really need two semaphores here? i think one would be sufficient to synchronize b/w two threads, please recheck once.

File: tests/kernel/threads/usage/src/task.c [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+	tick = _tick_get_32() - tick;
+
+	is_main_thread_ready = 1;
+	k_sem_take(&sem_thread1, K_FOREVER);
+
+	/*
+	 * By design this should be exact, but at least one cycle of
+	 * slop is required experimentally on Qemu.
+	 */
+	zassert_true(tick < SLEEP_TIME || tick > SLEEP_TIME + 1,
```
Comment (ramakrishnapallala) [inline tests/kernel/threads/usage/src/task.c]: you are comparing ticks with msec. you will have to do
__ticks_to_ms(tick)

also don't use  _tick_get_32() as it is kernel internal function, you can look into k_uptime_get_32() as it returns uptime in msec.

File: tests/kernel/threads/usage/src/task.c [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+	/*
+	 * Similarly check that thread2 ran for approximately
+	 * SLEEP_TIME. On QEMU, when the host CPU is overloaded, it
+	 * has been observed that the tick count can be missed by 1 on
+	 * either side. Allow for 2 ticks to be sure.  This check is
+	 * only there to make sure that thread2 did run for
+	 * approximately the whole time the main thread was sleeping.
+	 */
+	const int tick_error_allowed = 2;
```
Comment (ramakrishnapallala) [inline tests/kernel/threads/usage/src/task.c]: no need of variable definition here, the comment is good enough as it is anyway testing app.

File: tests/kernel/threads/usage/src/task.c [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+	TC_PRINT("Testing k_yield()\n");
+	thread_yield_test();
+
+	TC_PRINT("Testing k_thread_suspend() and k_thread_resume()\n");
+	thread_suspend_test();
+}
+
+/* The test to validate thread1_entry */
+void test_thread1(void)
```
Comment (ramakrishnapallala) [inline tests/kernel/threads/usage/src/task.c]: you don't need a separate test function, just move the k_thread_create logic to in one test functions as you are testing the only one thing by using two threads. so technically this is not a different test.

File: tests/kernel/threads/usage/src/task.c [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+	thread_macros_test(thread1_id, THREAD1_PRIO);
+
+	k_sem_give(&sem_thread2);
+	k_sem_take(&sem_thread1, K_FOREVER);
+
+	TC_PRINT("Testing k_thread_priority_set()\n");
+	thread_set_prio_test();
+
+	TC_PRINT("Testing k_sleep()\n");
+	thread_sleep_test();
```
Comment (ramakrishnapallala) [inline tests/kernel/threads/usage/src/task.c]: these can be your test cases.

File: tests/kernel/threads/usage/src/main.c (hunk 2 above)
Comment (ramakrishnapallala) [reply tests/kernel/threads/usage/src/main.c]: also do we need another file, we can just move the tests and test_main to a single file.

**Review by ramakrishnapallala — CHANGES_REQUESTED** [review]

File: tests/kernel/threads/usage/testcase.yaml (hunk 1 above)
Comment (nashif) [reply tests/kernel/threads/usage/testcase.yaml]: that is irrelevant,  please remove this from here and if there are issues, please document them in a comment.

**Review by andyross — COMMENTED** [review]
I'm not really understanding the purpose here.  All the APIs in question are (I think) already tested by the lifecycle test.  Can't you just add whatever scenarios you need there?

**Comment by rgundi** [comment]
@andyross : I had a detailed look at all the tests for threads and realized most of the tests (except for one particular use case in set_priority) have been addressed one way or the other (mainly in lifecycle tests and sleep tests). So, i modified the patchset to include only the set_priority tests.

@ramakrishnapallala : Your previous comments are not applicable for the new patchset now. Please review again.

File: tests/kernel/threads/lifecycle/lifecycle_api/src/test_threads_set_priority.c [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	zassert_equal(thread2_data, thread2_prio,
+		      "Expected priority to be changed to %d, not %d\n",
+		      thread2_prio, thread2_data);
+}
+
+
+/*test cases*/
+void test_threads_priority_set(void)
+{
+	threads_set_priority();
```
Comment (nashif) [inline tests/kernel/threads/lifecycle/lifecycle_api/src/test_threads_set_priority.c]: why not call this directly?

Comment (rgundi) [reply tests/kernel/threads/lifecycle/lifecycle_api/src/test_threads_set_priority.c]: My bad. modified it now.

**Review by ramakrishnapallala — APPROVED** [review]

**Pull request merged**